lib/web/recordingplayback.go

```diff
@@ -0,0 +1,767 @@
+		return nil, nil
+	}
+
+	playback := newRecordingPlayback(ctx, ws, clt, sessionID, h.logger)
+
+	playback.run()
+
+	return nil, nil
+}
+
+func (h *Handler) closeWebsocketWithError(ctx context.Context, ws *websocket.Conn, sessionID string, err error) {
+	data := []byte(err.Error())
+
+	totalSize := responseHeaderSize + len(data)
+	buf := make([]byte, totalSize)
+
+	encodeEvent(buf, 0, eventTypeError, 0, data, 0)
+
+	if err := ws.WriteMessage(websocket.BinaryMessage, buf); err != nil {
+		h.logger.ErrorContext(ctx, "failed to send event", "session_id", sessionID, "error", err)
+	}
+
+	deadline := time.Now().Add(websocketCloseTimeout)
+
+	// Send close frame to initiate graceful shutdown
+	if err := ws.SetWriteDeadline(deadline); err != nil {
+		h.logger.DebugContext(ctx, "failed to set write deadline", "session_id", sessionID, "error", err)
+	}
+	if err := ws.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")); err != nil {
+		h.logger.DebugContext(ctx, "failed to send close message", "session_id", sessionID, "error", err)
+	}
+
+	// Wait for peer's close frame response (or timeout)
+	if err := ws.SetReadDeadline(deadline); err != nil {
+		h.logger.DebugContext(ctx, "failed to set read deadline", "session_id", sessionID, "error", err)
+	}
+
+	// Log if we got something other than a close acknowledgement
+	if _, _, err := ws.ReadMessage(); err != nil && !websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
+		h.logger.DebugContext(ctx, "received non-close message while waiting for close acknowledgement",
+			"session_id", sessionID, "error", err)
+	}
+
+	// Finally close the underlying connection
+	ws.Close()
+}
+
+// newRecordingPlayback creates a new session recording playback handler.
+// This provides a way for the client to request session events within a specific time range, as well as the current
+// terminal screen state at a given time (when seeking).
+// This allows for faster seeking without having to send the client extra events to reconstruct the terminal state.
+func newRecordingPlayback(ctx context.Context, ws *websocket.Conn, clt events.SessionStreamer, sessionID string, logger *slog.Logger) *recordingPlayback {
+	ctx, cancel := context.WithCancel(ctx)
+
+	s := &recordingPlayback{
+		ctx:       ctx,
+		cancel:    cancel,
+		clt:       clt,
+		sessionID: sessionID,
+		logger:    logger,
+		ws:        ws,
+		writeChan: make(chan websocketMessage),
+	}
+
+	return s
+}
+
+// run starts the recording playback handler.
+func (s *recordingPlayback) run() {
+	defer s.cleanup()
+
+	go s.writeLoop()
+
+	s.readLoop()
+}
+
+// cleanup cleans up the recording playback resources.
+func (s *recordingPlayback) cleanup() {
+	s.cancel()
+
+	// Wait for any active task to complete
+	s.mu.Lock()
+	alreadySent := s.closeSent
+	s.mu.Unlock()
+
+	s.wg.Wait()
+
+	// Only send close message if we haven't already sent one
+	if !alreadySent {
```

Comment:
Does this is safe to assume that  alreadySent have up to data here ? It can be false if the `closeSent` was updated after the: 

```
	s.mu.Lock()
	alreadySent := s.closeSent
	s.mu.Unlock()
```
call.